src/telemetry.ts

```diff
@@ -36,4 +49,118 @@ export class NullTelemetryReporter implements ITelemetryReporter {
+    public constructor(private _telemetryReporter: TelemetryReporter, private _cadenceInMilliseconds: number = DefaultTelemetryIntervalInMilliseconds) {
+        this.reset();
+        this.setup();
+    }
+
+    public reportEvent(name: string, data?: any): void {
+        if (!this._eventBuckets[name]) {
+            this._eventBuckets[name] = [];
+        }
+
+        this._eventBuckets[name].push(data);
+    }
+
+    public finalize(): void {
+        this.send();
+        clearInterval(this._timer);
+    }
+
+    private setup(): void {
+        this._timer = setInterval(() => this.send(), this._cadenceInMilliseconds);
+    }
+
+    private reset(): void {
+        this._eventBuckets = {};
+    }
+
+    private send(): void {
+        for (const eventName in this._eventBuckets) {
+            const bucket = this._eventBuckets[eventName];
+            let properties = BatchTelemetryReporter.transfromBucketData(bucket);
+            this._telemetryReporter.reportEvent(eventName, properties);
+        }
+
+        this.reset();
+    }
+    /**
+     * Transfrom the bucket of events data from the form:
+     * [{
+     *  p1: v1,
+     *  p2: v2
+     * },
+     * {
+     *  p1: w1,
+     *  p2: w2
+     *  p3: w3
+     * }]
+     *
+     * to
+     * {
+     *   p1: [v1,   w1],
+     *   p2: [v2,   w2],
+     *   p3: [null, w3]
+     * }
+     *
+     *
+     * The later form is easier for downstream telemetry analysis.
+     */
+    private static transfromBucketData(bucket: any[]): {[groupedPropertyValue: string]: string} {
+        const allPropertyNamesInTheBucket = BatchTelemetryReporter.collectPropertyNamesFromAllEvents(bucket);
+        let properties = {};
+
+        // Run through all the entries, create a holder for all potential property names
+        for (const key of allPropertyNamesInTheBucket) {
+            properties[`aggregated.${key}`] = [];
+        }
+
+        // Run through again, collect the values for each property name.
+        for (const entry of bucket) {
+            for (const key of allPropertyNamesInTheBucket) {
+                properties[`aggregated.${key}`].push(entry[key] || null);
+            }
+        }
+
+        // Serialize each array as the final aggregated property value.
+        for (const key of allPropertyNamesInTheBucket) {
+            properties[`aggregated.${key}`] = JSON.stringify(properties[`aggregated.${key}`]);
+        }
+
+        return properties;
+    }
+
+    /**
+     * Get the property keys from all the entries of a event bucket:
+     *
+     * So
+     * [{
+     *  p1: v1,
+     *  p2: v2
+     * },
+     * {
+     *  p1: w1,
+     *  p2: w2
+     *  p3: w3
+     * }]
+     *
+     * will return ['p1', 'p2', 'p3']
+     */
+    private static collectPropertyNamesFromAllEvents(bucket: any[]): string[] {
+        let propertyNamesSet = {};
+        for (const entry of bucket) {
+            for (const key of Object.keys(entry)) {
+                propertyNamesSet[key] = true;
```

Comment:
What is the advantage of using a dictionary over an array for this?

Reply:
Array won't work. I am using dictionary as a set. Array can only append new keys (unless we search with `indexof()`). But a lot of entries will have the same  property names.

Reply:
That makes sense. I compleatly missed that :)

Could you add a comment saying that we use the dictionary to eliminate duplicated values?